### PR TITLE
Return key slice from QueueKeysFunc

### PR DESCRIPTION
**Pull request by qiujian16**

### CODE REVIEW

File: pkg/controller/factory/factory.go [hunk 1]
```diff
@@ -119,22 +124,54 @@ func (f *Factory) WithBareInformers(informers ...Informer) *Factory {
 // Pass the queueKeyFn you want to use to transform the informer runtime.Object into string key used by work queue.
 func (f *Factory) WithInformersQueueKeyFunc(queueKeyFn ObjectQueueKeyFunc, informers ...Informer) *Factory {
 	f.informerQueueKeys = append(f.informerQueueKeys, informersWithQueueKey{
-		informers:  informers,
-		queueKeyFn: queueKeyFn,
+		informers: informers,
+		queueKeysFn: func(obj runtime.Object) []string {
```
Comment (pmorie) [inline pkg/controller/factory/factory.go:128]: Do we want to change the existing method, or add a new one?

Comment (qiujian16) [reply pkg/controller/factory/factory.go:128]: I think this one is not exposed. We will still maintain the existing method in factory.

File: pkg/controller/factory/controller_context.go [hunk 2]
```diff
@@ -45,35 +45,47 @@ func (c syncContext) Recorder() events.Recorder {
...
+func (c syncContext) eventHandler(queueKeysFunc ObjectQueueKeysFunc, filter EventFilterFunc) cache.ResourceEventHandler {
 	resourceEventHandler := cache.ResourceEventHandlerFuncs{
 		AddFunc: func(obj interface{}) {
 			runtimeObj, ok := obj.(runtime.Object)
 			if !ok {
 				utilruntime.HandleError(fmt.Errorf("added object %+v is not runtime Object", obj))
 				return
 			}
-			c.Queue().Add(queueKeyFunc(runtimeObj))
+			keys := queueKeysFunc(runtimeObj)
```
Comment (mfojtik) [inline pkg/controller/factory/controller_context.go:56]:
```
for _, key := range queueKeyFunc(runtimeObj) {
  c.Queue().Add(key)
}
```

(you don't need keys variable)

File: pkg/controller/factory/factory.go [hunk 3]
```diff
@@ -68,6 +68,11 @@ type PostStartHook func(ctx context.Context, syncContext SyncContext) error
 // triggers.
 type ObjectQueueKeyFunc func(runtime.Object) string
```
Comment (mfojtik) [inline pkg/controller/factory/factory.go:69]: do we need this?

File: pkg/controller/factory/factory.go (hunk 1 above)
Comment (mfojtik) [reply pkg/controller/factory/factory.go:128]: I would refactor this to:

```
func WithInformersSingleQueueKeyFunc()
func WithInformersMultipleQueueKeysFunc()
```

File: pkg/controller/factory/factory.go [hunk 4]
```diff
@@ -119,22 +124,54 @@ func (f *Factory) WithBareInformers(informers ...Informer) *Factory {
 // Pass the queueKeyFn you want to use to transform the informer runtime.Object into string key used by work queue.
 func (f *Factory) WithInformersQueueKeyFunc(queueKeyFn ObjectQueueKeyFunc, informers ...Informer) *Factory {
 	f.informerQueueKeys = append(f.informerQueueKeys, informersWithQueueKey{
-		informers:  informers,
-		queueKeyFn: queueKeyFn,
+		informers: informers,
+		queueKeysFn: func(obj runtime.Object) []string {
+			key := queueKeyFn(obj)
+			return []string{key}
```
Comment (mfojtik) [inline pkg/controller/factory/factory.go:130]: just return `[]string{queueKeyFn(obj)}`

File: pkg/controller/factory/factory.go [hunk 5]
```diff
@@ -119,22 +124,54 @@ func (f *Factory) WithBareInformers(informers ...Informer) *Factory {
...
 }
 
 // WithFilteredEventsInformersQueueKeyFunc is used to register event handlers and get the caches synchronized functions.
 // Pass informers you want to use to react to changes on resources. If informer event is observed, then the Sync() function
 // is called.
-// Pass the queueKeyFn you want to use to transform the informer runtime.Object into string key used by work queue.
+// Pass the queueKeysFn you want to use to transform the informer runtime.Object into string key used by work queue.
 // Pass filter to filter out events that should not trigger Sync() call.
 func (f *Factory) WithFilteredEventsInformersQueueKeyFunc(queueKeyFn ObjectQueueKeyFunc, filter EventFilterFunc, informers ...Informer) *Factory {
 	f.informerQueueKeys = append(f.informerQueueKeys, informersWithQueueKey{
```
Comment (mfojtik) [inline pkg/controller/factory/factory.go:154]: s/informersWithQueueKey/informersWithQueueKeys/

File: pkg/controller/factory/factory.go [hunk 6]
```diff
@@ -119,22 +124,54 @@ func (f *Factory) WithBareInformers(informers ...Informer) *Factory {
...
+	})
+	return f
+}
+
+// WithFilteredEventsInformersQueueKeysFunc is used to register event handlers and get the caches synchronized functions.
+// Pass informers you want to use to react to changes on resources. If informer event is observed, then the Sync() function
+// is called.
+// Pass the queueKeysFn you want to use to transform the informer runtime.Object into a string slice of keys used by work queue.
+// Pass filter to filter out events that should not trigger Sync() call.
+func (f *Factory) WithFilteredEventsInformersQueueKeysFunc(queueKeysFn ObjectQueueKeysFunc, filter EventFilterFunc, informers ...Informer) *Factory {
```
Comment (mfojtik) [inline pkg/controller/factory/factory.go:170]: is there use-case for this?

**Comment by mfojtik** [comment]
/hold

Please hold merging this until we after master == 4.7.0

File: pkg/controller/factory/factory.go (hunk 1 above)
Comment (qiujian16) [reply pkg/controller/factory/factory.go:128]: change to WithInformersMultipleQueueKeysFunc(), should we keep the name of WithInformersQueueKeyFunc for compatible?

File: pkg/controller/factory/controller_context.go (hunk 2 above)
Comment (qiujian16) [reply pkg/controller/factory/controller_context.go:56]: done

File: pkg/controller/factory/factory.go (hunk 3 above)
Comment (qiujian16) [reply pkg/controller/factory/factory.go:69]: it is still needed as the parameter of WithInformersQueueKeyFunc

File: pkg/controller/factory/factory.go (hunk 5 above)
Comment (qiujian16) [reply pkg/controller/factory/factory.go:154]: this is still to use a func that returns single key

File: pkg/controller/factory/factory.go (hunk 4 above)
Comment (qiujian16) [reply pkg/controller/factory/factory.go:130]: done

File: pkg/controller/factory/factory.go (hunk 6 above)
Comment (qiujian16) [reply pkg/controller/factory/factory.go]: currently no,  i will remove this func

**Pull request closed without merging**